### PR TITLE
Improve migration system vol2

### LINKED ISSUE #216: Migration Improvments Vol 2
- [x] Page for database list with links for each database's migrations with a checkbox to show only the failed migrations.
- [x] Add table filter to update schema on all databases
- [x] Performance improvments

### PR DESCRIPTION
**Pull request by Baxxx**
Closes #216 

Perfomance sheet for average computing time for **_n**_ iterations or loading all migrations for 4 databases, 3 samples for both 1k and 5k iterations for both versions.

| n | BEFORE [ms] | AFTER [ms] | FACTOR [BEFORE/AFTER] |
| --- | --- | --- | --- |
| 1000 | 4.750 | 1.794 | 2.65 |
| 1000 | 4.702 | 1.627 | 2.89 |
| 1000 | 4.718 | 1.636 | 2.88 |
| 5000 | 4.758 | 1.589 | 2.99 |
| 5000 | 4.676 | 1.572 | 2.97 |
| 5000 | 4.715 | 1.581 | 2.98 |

### CODE REVIEW

**Comment by sls-lnluksa** [comment]
**Recommendation:**
Add setup eg. BROWSER and CONSOLE which will have different new line separators, colors, and display data a bit differently. 
Especially color red for failed queries log. If we have a lot databases, it is easier to spot failed one.

**Comment by sls-lnluksa** [comment]
First round review: Nikola Lukovic (sls-lnluksa)
Final review with: @stoivo

**Comment by Baxxx** [comment]
You should be able to display all failed migrations immediately from list page.

File: modules/migration/model/migration.class.php [hunk 1]
```diff
@@ -38,42 +43,37 @@ function get_migrations_for_database($db_name) {
...
+
+    foreach ($dbs as $db) {
+      $all_migrations[$db->Database] = $this->get_migrations_for_database($db->Database);
     }
 
     return $all_migrations;
   }
 
-  function get_migrations_for_all_databases() {
+  function get_database_names() {
```
Comment (Baxxx) [inline modules/migration/model/migration.class.php]: Do not copy the method. Use an argument instead.

File: modules/migration/view/list.php [hunk 2]
```diff
@@ -20,58 +20,80 @@
...
+<? if(!$show_only_database_names) { ?>
+<form action="<? print $current_page ?>" method="post">
   <input type="submit" name="action_update_schema" value="Update schema information on all DBs"/>
+  Table filter: <input type="text" name="migration_tablefilter">
 </form>
+<? } ?>
 <?
 foreach ($all_migrations as $database => $migrations) {
-  print "<h2>". $database ."</h2>";
+  print "<h2><a href='". $MY_SELF ."db_name=". $database."'>". $database ."</a></h2>";
```
Comment (Baxxx) [inline modules/migration/view/list.php]: Pass `show_only_failed` argument as well.

File: modules/migration/view/list.php [hunk 3]
```diff
@@ -42,6 +42,10 @@
      - <a href="<? print $MY_SELF . ($show_all ? "show_all=1&" : "") . ($db_name_restriction ? "db_name=".$db_name_restriction."&" : "") ?>">Show all migrations</a><br>  
   <? } ?>
    - <a href="<? print $MY_SELF ?>">Back to list</a>
+<? } else { ?>
+  <? if(!$show_only_failed) { ?>
+     - <a href="<? print $MY_SELF . "show_all=1&" . ($db_name_restriction ? "db_name=".$db_name_restriction."&" : "") ?>show_only_failed=1">Show only failed migrations for all databases</a><br>
```
Comment (Baxxx) [inline modules/migration/view/list.php]: Remove `db_name` part.

**Pull request merged**